olp-cpp-sdk-dataservice-read/src/StreamLayerClientImpl.cpp

```diff
@@ -21,23 +21,142 @@
-      settings_(std::move(settings)) {
+      settings_(std::move(settings)),
+      pending_requests_(std::make_shared<client::PendingRequests>()) {
   if (!settings_.cache) {
     settings_.cache = client::OlpClientSettingsFactory::CreateDefaultCache({});
   }
 }
 
-StreamLayerClientImpl::~StreamLayerClientImpl() = default;
+StreamLayerClientImpl::~StreamLayerClientImpl() {
+  pending_requests_->CancelAllAndWait();
+}
+
+bool StreamLayerClientImpl::CancelPendingRequests() {
+  OLP_SDK_LOG_TRACE(kLogTag, "CancelPendingRequests");
+  return pending_requests_->CancelAll();
+}
+
+client::CancellationToken StreamLayerClientImpl::Subscribe(
+    SubscribeRequest request,
+    SubscribeRequest::SubscribeResponseCallback callback) {
+  using SubscribeResponse =
+      client::ApiResponse<model::SubscribeResponse, client::ApiError>;
+
+  auto self = shared_from_this();
+
+  {
+    std::unique_lock<std::mutex> lock(mutex_);
+    if (subscription_id_) {
+      lock.unlock();
+      OLP_SDK_LOG_WARNING_F(kLogTag, "Subscribe : already subscribed");
+
+      callback(client::ApiError(client::ErrorCode::InvalidArgument,
+                                "Already subscribed", false));
+      return client::CancellationToken();
+    }
+  }
+
+  const auto& subscription_id = request.GetSubscriptionId();
+  const auto subscription_mode =
+      GetSubscriptionMode(request.GetSubscriptionMode());
+  const auto& consumer_id = request.GetConsumerId();
+  const auto& properties = request.GetConsumerProperties();
+  std::string correlation_id;
+
+  auto subscribe_task =
+      [=, &correlation_id](
+          client::CancellationContext context) -> SubscribeResponse {
+    auto stream_api = ApiClientLookup::LookupApi(
+        catalog_, context, kStreamService, kStreamVersion,
+        FetchOptions::OnlineIfNotFound, settings_);
+
+    if (!stream_api.IsSuccessful()) {
+      return stream_api.GetError();
+    }
+
+    const auto subscription = StreamApi::Subscribe(
+        stream_api.GetResult(), layer_id_, subscription_id, subscription_mode,
+        consumer_id, properties, context, correlation_id);
+
+    if (!subscription.IsSuccessful()) {
+      return subscription.GetError();
+    }
+
+    return subscription;
+  };
+
+  return AddTask(
+      settings_.task_scheduler, pending_requests_, std::move(subscribe_task),
+      [=](SubscribeResponse response) {
+        if (response.IsSuccessful()) {
+          const auto subscription_id = response.GetResult().GetSubscriptionId();
+          const auto node_base_url = response.GetResult().GetNodeBaseURL();
+
+          {
+            std::lock_guard<std::mutex> lock(self->mutex_);
+            self->subscription_id_ = subscription_id;
+            self->subscription_mode_ = subscription_mode;
+            self->x_correlation_id = correlation_id;
+            self->node_base_url_ = node_base_url;
+          }
+
+          callback(subscription_id);
+        } else {
+          callback(response.GetError());
```

Comment:
Do you clear all the previous cached subscription fields (see `subscription_id_`, `subscription_mode_`, etc.) in case of error?